### PR TITLE
feat: add support for storing and validating GBFS files in GCP storage

### LINKED ISSUE #566: Support GBFS Feeds with Validation Reports
#### Description:
To enhance our support for GBFS feeds, we need to implement a process to run GBFS validation reports using the `gbfs-validator`. This involves scheduling validation processes, designing appropriate database entities, and ensuring only updated feeds and their validation reports are stored.

#### Tasks:
- [x] **Schedule Validation Process:** Implement a mechanism to trigger the validation process for each GBFS feed on a defined schedule.
- [x] **Design Database Entities:** Design and implement database entities to store the validation reports.
- [x] **Store Updated Feeds and Reports:** Ensure that GBFS feeds and their related validation reports are stored.

#### Dependency:
- Requires https://github.com/MobilityData/gbfs-validator/pull/101 to be merged to prevent excessive HTTP response sizes.

### PR DESCRIPTION
**Pull request by cka-y**
**Summary:**

Closes #566 by adding support for storing GBFS files from an auto-discovery URL in a Cloud Storage bucket. The updates include the creation of a new `gbfs.json` file with paths pointing to the stored files in the bucket, validating the feeds, and storing the necessary information in the database. The changes involve the creation of the `gbfs-validator-batch` and `gbfs-validator-pubsub` functions to support this new functionality, ensuring that GBFS feeds are properly validated and stored.

**Expected behavior:**

This pull request introduces the following behavior:
- The `gbfs-validator-batch` function triggers an execution that checks all GBFS feeds in the database and publishes a message to the Pub/Sub topic for each feed.
- The `gbfs-validator-pubsub` function processes these messages, downloads feed snapshots to GCP, validates them, saves the validation report summary to GCP, and updates the database with the snapshot information and validation report.

**Testing tips:**

1. Trigger the `batch` function:
   - Execute the following command to trigger the batch function:
     ```bash
     curl -m 550 -X POST https://northamerica-northeast1-mobility-feeds-dev.cloudfunctions.net/gbfs-validator-batch \
     -H "Authorization: bearer $(gcloud auth print-identity-token)" \
     -H "Content-Type: application/json"
     ```
   - Alternatively, trigger the Cloud Scheduler by selecting [`gbfs-validator-batch-scheduler-dev`](https://console.cloud.google.com/cloudscheduler?project=mobility-feeds-dev) and clicking `Force Run`.

2. The `gbfs-validator-pubsub` function should be triggered automatically. You can verify the logs to ensure everything is functioning correctly: [Pub/Sub Function Logs](https://console.cloud.google.com/functions/details/northamerica-northeast1/gbfs-validator-pubsub?env=gen2&project=mobility-feeds-dev&tab=logs).

3. Check the stored feeds in the specified Cloud Storage bucket. Ensure the `gbfs.json` file and the validation report are correctly generated: [GBFS Storage Bucket](https://console.cloud.google.com/storage/browser/mobilitydata-gbfs-snapshots-dev;tab=objects?forceOnBucketsSortingFiltering=true&project=mobility-feeds-dev&prefix=&forceOnObjectsSortingFiltering=false).

4. Verify that the database content in the DEV env reflects the expected changes and updates.

Testers are invited to follow the tips AND to try anything they deem relevant outside the bounds of the testing tips. 

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the unit tests with `./scripts/api-tests.sh` to make sure you didn't break anything
- [ ] Add or update any needed documentation to the repo
- [ ] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

**Comment by cka-y** [comment]
The integration tests are expected to fail because of the database changes

File: functions-python/gbfs_validator/src/gbfs_utils.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                new_url = upload_gbfs_file_to_bucket(bucket, old_url, blob_name)
+                feed_info["url"] = new_url
+        elif isinstance(feed["feeds"], list):
+            for feed_info in feed["feeds"]:
+                old_url = feed_info["url"]
+                blob_name = f"{stable_id}/{stable_id}-{today}/{feed_info['name']}.json"
+                new_url = upload_gbfs_file_to_bucket(bucket, old_url, blob_name)
+                feed_info["url"] = new_url
+        else:
+            logging.warning(f"Unexpected format in feed: {feed_key}")
```
Comment (davidgamez) [inline functions-python/gbfs_validator/src/gbfs_utils.py:73]: [suggestion]: I would suggest adding a try/catch section to get a detailed error message if one of the URLs fails while downloading.

File: functions-python/gbfs_validator/src/gbfs_utils.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    trace_service.save(trace)
+
+
+def create_snapshot(stable_id: str, feed_id: str, hosted_url: str) -> Gbfssnapshot:
+    """Create a new Gbfssnapshot object."""
+    today = datetime.now().strftime("%Y-%m-%d")
+    snapshot_id = str(uuid.uuid4())
+    snapshot = Gbfssnapshot(
+        id=snapshot_id,
+        stable_id=f"{stable_id}-{today}",
```
Comment (davidgamez) [inline functions-python/gbfs_validator/src/gbfs_utils.py:98]: [question]: Does this mean that we cannot have more than one snapshot per day?

File: functions-python/gbfs_validator/src/gbfs_utils.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+def save_trace_with_error(trace, error, trace_service):
+    """Helper function to save trace with an error."""
+    trace.error_message = error
+    trace.status = Status.FAILED
+    trace_service.save(trace)
+
+
+def create_snapshot(stable_id: str, feed_id: str, hosted_url: str) -> Gbfssnapshot:
+    """Create a new Gbfssnapshot object."""
+    today = datetime.now().strftime("%Y-%m-%d")
```
Comment (davidgamez) [inline functions-python/gbfs_validator/src/gbfs_utils.py:94]: I have seen this multiple times across the code. All points to the creation of a snapshot ID. We should centralize the creation of the ID to make sure if the ID pattern changes, there is only one place to change.

File: functions-python/gbfs_validator/src/gbfs_utils.py (hunk 2 above)
Comment (cka-y) [reply functions-python/gbfs_validator/src/gbfs_utils.py:98]: Yes. Would you suggest making it like the gtfs dataset (including timestamp to the minute)?

Comment (davidgamez) [reply functions-python/gbfs_validator/src/gbfs_utils.py:98]: Yes, to-the-minute accuracy will be aligned with the  GTFS dataset.

File: functions-python/gbfs_validator/src/gbfs_utils.py (hunk 3 above)
Comment (cka-y) [reply functions-python/gbfs_validator/src/gbfs_utils.py]: good catch! i have created a `get_snapshot_id` function that is called from all function that constructed the id

File: functions-python/gbfs_validator/src/gbfs_utils.py (hunk 2 above)
Comment (cka-y) [reply functions-python/gbfs_validator/src/gbfs_utils.py]: I created a new `GBFSValidator` class thanks to your suggestion

**Review by davidgamez — APPROVED** [review]
LGTM!

**Pull request merged**